### PR TITLE
ELPP-2463 Web test cases

### PR DESCRIPTION
**Pull request by stephenwf**
I've added an example test case that I will split out into its own. This gives 3 examples of inserting and retrieving. This functionality is meant to test the user-facing components, the parts excluded are:

- Creating relations based on rules
- Caching logic for `get(type, id)`
- All API-SDK integrations and fetching

We are using the `Builder` class from the SDK to build the fixtures to test from. The API is open enough to use all the models that recommendations supports, and the relations are explicit and can be tested in isolation. Each test at the moment reads quite well, with the intentions of whats being tested clear.

### CODE REVIEW

File: src/Recommendations/Command/GenerateDatabaseCommand.php [hunk 1]
```diff
@@ -48,7 +54,7 @@ private function tableExists(string $table): bool
 
     private function createTables(bool $drop, Table ...$tables)
     {
-        foreach ($tables as $table) {
+        foreach (array_reverse($tables) as $table) {
```
Comment (stephenwf) [inline src/Recommendations/Command/GenerateDatabaseCommand.php]: @giorgiosironi This solves the database creation/deletion issues. This way we delete the tables in the reverse order they are created

**Comment by thewilkybarkid** [comment]
Could we do the testing at a higher-level, actually mocking the API and notifications rather than creating the rules? That way it would guarantee that a particular dataset produces a particular result, without knowledge of the internals?

File: src/App/Kernel.php [hunk 2]
```diff
@@ -72,7 +73,12 @@
     private $startTime;
     private $version;
 
-    public function __construct($config = [])
+    public function getApp()
```
Comment (giorgiosironi) [inline src/App/Kernel.php]: constructor should always be the first method, cosmetic change

File: src/Recommendations/RuleModelRepository.php [hunk 3]
```diff
@@ -37,14 +39,14 @@ public function getLatestArticle()
...
-          SELECT * FROM Rules 
-          WHERE Rules.type ="research-advance"
-          OR Rules.type ="research-article"
-          OR Rules.type ="research-exchange"
-          OR Rules.type ="short-report"
-          OR Rules.type ="tools-resources"
-          OR Rules.type ="replication-study" 
-          ORDER BY Rules.published DESC
+          SELECT * FROM '.$this->ruleTableName.' as Ru 
+          WHERE Ru.type ="research-advance"
```
Comment (giorgiosironi) [inline src/Recommendations/RuleModelRepository.php]: is this supposed to be used only in testing or also production? By the way, `WHERE RU.type IN (...)` might be faster due to index usage?

File: tests/src/Web/WebTestCase.php [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+        parent::setUp();
+        $lines = $this->runCommand('generate:database');
+//        $this->assertStringStartsWith('Database created successfully.', $lines[0], 'Failed to run test during set up');
+    }
+
+    public function tearDown()
+    {
+        $this->getDatabase()->query(sprintf('SET FOREIGN_KEY_CHECKS=%s', (int) false));
+        // Delete database.
+        $this->getDatabase()->getSchemaManager()->dropTable('References__test');
```
Comment (giorgiosironi) [inline tests/src/Web/WebTestCase.php]: could be optimized to only delete data rather than dropping the tables, if the tests take a long time to run

Comment (stephenwf) [inline tests/src/Web/WebTestCase.php]: Thats true, although the `setUp` would have to have an `if table exists drop` type thing, which has proven to be painful!

**Comment by stephenwf** [comment]
@giorgiosironi I'm unsure why that last build failed, it said missing composer, but I can't see any errors

**Comment by giorgiosironi** [comment]
The build was failing because it was requiring Composer 1.0 as part of the Puli setup, we removed that recently. Updating and rerunning now.

**Comment by giorgiosironi** [comment]
Ok, now it fails on the database creation.

**Comment by stephenwf** [comment]
@giorgiosironi `14:33:26 NetworkError: Timed out trying to connect to 54.82.238.196 (tried 1 time)`

File: composer.json [hunk 5]
```diff
@@ -27,7 +27,8 @@
         "elife/logging-sdk": "dev-master",
         "ocramius/package-versions": "^1.1",
         "justinrainbow/json-schema": "^5.1",
-        "mindplay/composer-locator": "^2.1"
+        "mindplay/composer-locator": "^2.1",
+        "behat/gherkin": "^4.4"
```
Comment (thewilkybarkid) [inline composer.json]: Not used?

**Comment by thewilkybarkid** [comment]
@giorgiosironi Tests seem to pass, but then it's a `Parallel execution exception`.

**Comment by stephenwf** [comment]
`Total rules in the index: 0` maybe this is the issue

**Comment by giorgiosironi** [comment]
Yes, selected relevant output:
```
09:04:36 [54.174.222.171] out: Job in SQS queue (approximate): 0
09:04:36 [54.174.222.171] out: No more SQS messages, finished waiting
09:04:36 [54.174.222.171] out: Total rules in the index: 0
09:04:36 [54.174.222.171] out: There s
09:04:36 Fatal error: run() received nonzero return code 2 while executing!
...
09:04:36 hould be exactly 29 in the database
09:04:36 [54.174.222.171] out: 
```

**Pull request merged**